packages/vite/src/node/build.ts

```diff
@@ -411,7 +411,9 @@ async function doBuild(
     } catch (e) {}
     if (!knownImports) {
       // no dev deps optimization data, do a fresh scan
-      knownImports = Object.keys((await scanImports(config)).deps)
+      const deps = (await scanImports(config)).deps
+      await addManuallyIncludedOptimizeDeps(deps, config)
+      knownImports = Object.keys(deps)
```

Comment:
I think it would be nice if `scanImports` handle the manually included optimize deps too so we can prevent mutating `deps`, and have one function to do it all.  We would need to refactor a bit of how we report missing deps in the optimizer, but I think it looks a bit cleaner. Curious on your thoughts or caveats you see.

Reply:
I thought about this, and it was the first thing I tried. But I don't know if we should start throwing on missing imports during build, so we should keep that logic in the optimizer instead of the scan phase. What we could do is to create a new function in the optimizer:
```ts
export function findKnownImports(config: ResolvedConfig) { 
  const deps = (await scanImports(config)).deps
  await addManuallyIncludedOptimizeDeps(deps, config)
  return Object.keys(deps)
}
```
So we simplify the code on build.ts, and we avoid exposing the internal `addManuallyIncludedOptimizeDeps` function. What do you think?

Reply:
I think that's fine, but I was hoping we could refactor [this part](https://github.com/vitejs/vite/blob/d1ec9d258a42bb460932d7a951a1426f893e2482/packages/vite/src/node/optimizer/index.ts#L258-L281) together as well. E.g. `scanImports` would return `{ deps, missingIds, missingOptimizeDeps }` and we log it out later. So we keep `scanImports` as the sole function for scanning. I'm happy to merge this too and explore this refactor later though if you prefer.

Reply:
Im afraid of throwing a new exception during build. Lets merge this one and maybe you can later do a refactor PR so we check how it would work. We could be more bold for 3.0 also